### ISSUE TITLE
[Bug Report] Bug in `get_caching_hooks` when `pos_slice=None`

### ISSUE THREAD

**Issue by hannamw** [neelnanda-io/TransformerLens#574]
**Describe the bug**
If you call `get_caching_hooks` without specifying `pos_slice` (or if you set `pos_slice=None`), and then run the model with these hooks, an `assert` checking that `pos_slice is not None` will be triggered. Looking at how the `pos_slice` is handled in `run_with_cache`, I believe that the correct solution here is to add in a couple lines that say 
```
if not isinstance(pos_slice, Slice):
    if isinstance(
        pos_slice, int
    ):  # slicing with an int collapses the dimension so this stops the pos dimension from collapsing
        pos_slice = [pos_slice]
    pos_slice = Slice(pos_slice)
```
It might also just make sense to move those lines into `get_caching_hooks`. Maybe this is user error on my end—I've been using `get_caching_hooks` and `with model.hooks()` to cache and intervene during the same forward pass, but let me know if I shouldn't be using `get_caching_hooks` for this at all.

**Code example**
```
from transformer_lens import HookedTransformer

model = HookedTransformer.from_pretrained('gpt2')
s = "This is a test sentence."

# This works
_, cache = model.run_with_cache(s)

# This doesn't
cache, hooks, _ = model.get_caching_hooks()
with model.hooks(hooks):
    model(s)
```

This code will break at line 545 (`assert pos_slice is not None  # keep mypy happy`) in `hook_points.py`.

**System Info**
Describe the characteristic of your environment:
 * Describe how `transformer_lens` was installed (pip, docker, source, ...): **pip (1.17.0)**
 * What OS are you using? (Linux, MacOS, Windows): **Linux**
 * Python version (We suppourt 3.7 -3.10 currently): **Python 3.10**

**Additional context**

### Checklist

- [x] I have checked that there is no similar [issue](https://github.com/neelnanda-io/TransformerLens/issues) in the repo (**required**)

**Comment by janraasch** [comment]
Hi @hannamw,

I ran into the same issue when trying to run the "Get activations"-section 

```python
...
    with model.hooks(fwd_hooks=caching_hooks):
        _ = model(prompt)
    return cache[name]
...
```

of https://colab.research.google.com/drive/1x5EmvAdVmmHvOOd1jxvODj233wbvbl1U#scrollTo=2h6X8zdojKHN which is a notebook referenced in »Activation Addition: Steering Language Models Without Optimization« (see https://arxiv.org/pdf/2308.10248).

**hunch** I think the offending/problematic commit/line seems to be https://github.com/neelnanda-io/TransformerLens/pull/516/files#diff-85fafabc38ed61e857e7db5df426fe1696f7a1dbe3a9c532ca8dff4718d0193dR566.

ℹ️ It's kinda my first day at work here, so I am not sure how to best fix this 😇. I'm just getting started with `mech interp` as part of my [AI Safety Fundamentals Project](https://aisafetyfundamentals.com/projects/).

**Comment by hannamw** [comment]
Thanks for confirming you're seeing this issue too; it's good to know that I'm not the only one!

With respect to the cause of this issue, that assertion is indeed the direct cause of the failure. That said, even if it is removed, the code will fail, as the line below (`resid_stream = pos_slice.apply(resid_stream, dim=pos_dim)`) assumes that `pos_slice` has an `apply` method; if `pos_slice` is `None`, it will not.

I can make a PR soon to fix this issue.

**Comment by hannamw** [comment]
Fixed by #578
